build-logic-settings/testfiltering-plugin/src/main/kotlin/gradlebuild.internal.testfiltering.settings.gradle.kts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+val branchFuture = CompletableFuture<String>()
+Thread {
+    val cmd = arrayOf("git", "rev-parse", "--abbrev-ref", "HEAD")
+    try {
+        val proc = ProcessBuilder(*cmd).directory(rootDir).redirectErrorStream(true).start()
+        if (proc.waitFor(10, TimeUnit.SECONDS)) {
+            val text = proc.inputStream.bufferedReader().readText().trimEnd()
+            val exitValue = proc.exitValue()
+            if (exitValue == 0) {
+                branchFuture.complete(text)
+            } else {
+                throw IllegalStateException("Command [${cmd.joinToString(" ")}] failed with $exitValue: $text")
+            }
+        } else {
+            proc.destroy()
+            throw IllegalStateException("Command [${cmd.joinToString(" ")}] timed out")
+        }
+    } catch (any: Exception) {
+        branchFuture.completeExceptionally(any)
+    }
+}.start()
+
+val testFilteringProperty = "gradle.internal.testselection.enabled"
+val gitBranchName: String? = branchFuture.get()
+if (!System.getProperties().containsKey(testFilteringProperty) && "CI" in System.getenv() && gitBranchName != null) {
```

Comment:
Why is this whole logic here? Can't this go into the individual projects? The we could re-use the already determined git branch.

I think the configuration should use a system property which we set on the daemon JVM but use some DSL on the test task/the distributed testing extension instead. Is that already possible?